assets/src/data/model/entity-factory/extractors.js

```diff
@@ -0,0 +1,348 @@
+ * provided entity instance
+ * @param {Object} entityInstance
+ * @return {Object} A plain object
+ */
+export const getBaseFieldsAndValuesForPersisting = ( entityInstance ) => {
+	return reduce( entityInstance.originalFieldsAndValues, (
+		fieldsAndValues,
+		fieldValue,
+		fieldName
+	) => {
+		if (
+			isEntityField( fieldName, entityInstance.schema ) &&
+			! isPrimaryKeyField( fieldName, entityInstance.schema )
+		) {
+			return fieldsAndValues[ fieldName ] = maybeConvertFromValueObject(
+				fieldValue,
+			);
+		}
+	}, {} );
+};
+
+/**
+ * Returns the primary key(s) and values for the given entityInstance
+ *
+ * @param {Object} entityInstance
+ * @return {Object} an array of values for the primary keys.
+ */
+export const getPrimaryKeyValues = ( entityInstance ) => pick(
+	entityInstance,
+	entityInstance.primaryKeys
+);
+
+/**
+ * This returns a plain object of entity fields from the schema for the entity
+ * instance (schema for fields are extracted as well).
+ *
+ * @param {Object} entityInstance
+ * @return {Object} A plain object with fields and schema properties that are
+ * entity properties.
+ */
+export const getEntityFieldsFromSchema = ( entityInstance ) => pickBy(
+	entityInstance.schema,
+	( fieldValue, fieldName ) => isEntityField(
+		fieldName,
+		entityInstance.schema
+	)
+);
+
+/**
+ * This returns a plain object of extracted primaryKey fields from the schema
+ * for the entity instance.
+ *
+ * @param {Object} entityInstance
+ * @return {Object} A plain object with fields and schema properties that
+ * 					represent primary key fields.
+ */
+export const getPrimaryKeyFieldsFromSchema = ( entityInstance ) => pickBy(
+	entityInstance.schema,
+	( fieldValue, fieldName ) => isPrimaryKeyField(
+		fieldName,
+		entityInstance.schema
+	)
+);
+
+/**
+ * Derives the default value to use for a given type.
+ *
+ * @param {string} type
+ * @return {*}  A value to use for the given type.
+ */
+export const deriveDefaultValueForType = ( type ) => {
+	if ( isArray( type ) ) {
+		return type.indexOf( 'null' ) > -1 ?
+			null :
+			deriveDefaultValueForType( type[ 0 ] );
+	}
+	switch ( type ) {
+		case 'string':
+			return '';
+		case 'number':
+		case 'integer':
+			return 0;
+		case 'null':
+		case 'object':
+			return null;
+		case 'boolean':
+		case 'bool':
+			return false;
+		case 'date-time':
+			return ( new Date() ).toISOString();
+	}
+	return null;
+};
+
+/**
+ * Derives what `type` a field is from the schema.
+ * It accounts for cases where the "type" of a field might be `date-time` or
+ * where the type is an object and thus the `type` for the purposes of model
+ * entities is defined by the `raw` property for the field.
+ *
+ * @param {string} fieldName
+ * @param {Object} schema
+ * @return {*}  What type the filed is.
+ */
+export const deriveTypeForField = ( fieldName, schema ) => {
+	if ( isDateTimeField( fieldName, schema ) ) {
+		return 'date-time';
+	}
```

Comment:
do we not need to detect money fields here as well?

should this maybe also use something like the `ValueObjectExtratorCollection` from above?